stonesoup/hypothesiser/composite.py

```diff
@@ -0,0 +1,148 @@
+from ..predictor.composite import CompositePredictor
+from ..types.detection import MissedDetection, CompositeMissedDetection
+from ..types.hypothesis import CompositeProbabilityHypothesis
+from ..types.multihypothesis import MultipleHypothesis
+
+
+class CompositeHypothesiser(Hypothesiser):
+    """A composition of multiple sub-hypothesisers.
+
+    Generate track predictions at detection times and calculate probabilities for all
+    prediction-detection pairs for single prediction and multiple detections.
+
+    Assumes tracks and measurements exist in composite state spaces.
+    Utilises the output hypotheses of its :attr:`sub_hypothesisers` to attain a
+    :class:`~.CompositeHypothesis` for each prediction-detection pair.
+    """
+    sub_hypothesisers: Sequence[Hypothesiser] = Property(
+        doc="A sequence of sub-hypothesisers. These must be hypothesisers that return "
+            "probability-weighted hypotheses, in order for composite hypothesis weights to be "
+            "calculated.")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # create predictor as composition of sub-hypothesisers' predictors
+        sub_predictors = list()
+        for sub_hypothesiser in self.sub_hypothesisers:
+            sub_predictors.append(sub_hypothesiser.predictor)
+        self.predictor = CompositePredictor(sub_predictors)
+
+    def hypothesise(self, track, detections, timestamp):
+        """ Evaluate and return all track association hypotheses.
+
+        For a given track and a set of N available detections, return a MultipleHypothesis object
+        with N+1 detections (first detection is a 'CompositeMissedDetection'), each with an
+        associated probability.
+
+        A resultant composite hypothesis' probability is calculated as the weighted product of
+
+        Parameters
+        ----------
+        track: :class:`~.Track`
+            The track object to hypothesise on, existing in a composite state space
+        detections: :class:`set`
+            A set of :class:`~CompositeDetection` objects, representing the available detections.
+        timestamp: :class:`datetime.datetime`
+            A timestamp used when evaluating the state and measurement predictions. Note that if a
+            given detection has a non empty timestamp, then prediction will be performed according
+            to the timestamp of the detection.
+
+        Returns
+        -------
+        : :class:`~.MultipleHypothesis`
+            A container of :class:`~CompositeHypothesis` objects, each of which containing a
+            sequence of :class:`~.SingleHypothesis` objects
+
+        Notes
+        ----
+        * Detections are required to have a timestamp, in order for the composite null-hypothesis
+        to have a timestamp.
+        * A 'weightings' attribute should be implemented to adjust sub-hypotheses' contributions
+        to the composite hypothesis weights.
```

Comment:
Warning from docs build:
```
stonesoup/hypothesiser/composite.py:docstring of stonesoup.hypothesiser.composite.CompositeHypothesiser.hypothesise:25: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

Reply:
I think this is now fixed. Though there is now another docs build issue.